File: killer_sudoku/solver/engine/rules/naked_single.py
"""R1a NakedSingle — recognise a cell reduced to a single candidate.

Fires on CELL_DETERMINED.  Returns no eliminations — the engine has
already reduced candidates[r][c] to a singleton.  This rule exists as
a named concept so the coaching layer can generate a placement hint:
"cell (r,c) has only one remaining candidate (d) — place d there."

Peer eliminations that follow placement are handled by CellSolutionElimination
(R1b), which fires on CELL_SOLVED after NakedSingle acknowledges the cell.
"""

from __future__ import annotations

from killer_sudoku.solver.engine.hint import HintResult
from killer_sudoku.solver.engine.rule import RuleContext
from killer_sudoku.solver.engine.rules._registry import hintable_rule
from killer_sudoku.solver.engine.types import (
    Elimination,
    Placement,
    RuleResult,
    Trigger,
    UnitKind,
)


@hintable_rule
class NakedSingle:
    """R1a: named recognition of a cell determined by a single candidate."""

    name = "NakedSingle"
    description = (
        "When a cell has only one remaining candidate, that digit must go there. Also "
        "removes it from peer cells in the same row, column, and box."
    )
    priority = 0
    triggers: frozenset[Trigger] = frozenset({Trigger.CELL_DETERMINED})
    unit_kinds: frozenset[UnitKind] = frozenset()  # cell-scoped

    def apply(self, ctx: RuleContext) -> RuleResult:
        """Return a placement for the determined cell.

        When NakedSingle is always-apply, the engine promotes the placement to
        user_grid automatically.  When hint-only, as_hints() surfaces it as a
        coaching suggestion instead.
        """
        assert ctx.cell is not None
        assert ctx.hint_digit is not None
        return RuleResult(placements=[Placement(cell=ctx.cell, digit=ctx.hint_digit)])

    def as_hints(
        self, ctx: RuleContext, eliminations: list[Elimination]
    ) -> list[HintResult]:
        """Return a placement hint for the determined cell.

        NakedSingle fires on CELL_DETERMINED (ctx.cell set, ctx.hint_digit set).
        The placement hint instructs the user to place the digit — no eliminations.
        """
        assert ctx.cell is not None
        assert ctx.hint_digit is not None
        r, c = ctx.cell
        d = ctx.hint_digit
        return [
            HintResult(
                rule_name=self.name,
                display_name="Naked Single",
                explanation=(
                    f"Cell r{r + 1}c{c + 1} has only one remaining candidate:"
                    f" {d}. Place {d} there."
                ),
                highlight_cells=frozenset({(r, c)}),
                eliminations=[],
                placement=(r, c, d),
            )
        ]

File: web/src/engine/rules/nakedSingle.ts
/**
 * NakedSingle — R1: cell with a single candidate receives that digit, and that
 * digit is removed from all row/col/box/distinct-cage peers.
 *
 * Combines the former NakedSingle (placement) and CellSolutionElimination (peer
 * elimination) into a single rule so both effects are always applied together.
 *
 * Fires on CELL_DETERMINED (ctx.cell and ctx.hintDigit set by the engine when
 * a candidate set collapses to a singleton). Returns a Placement and all peer
 * Eliminations in one RuleResult.
 */

import type { HintResult } from '../hint.js';
import type { RuleContext } from '../rule.js';
import {
  Cell,
  Elimination,
  emptyResult,
  RuleResult,
  Trigger,
  UnitKind,
} from '../types.js';
import { cellLabel } from './_labels.js';

export class NakedSingle {
  readonly name = 'NakedSingle';
  readonly displayName = 'Naked Single';
  readonly description = `\
Naked Single — a cell reduced to one candidate must hold that digit, and that digit
is eliminated from all peers in shared rows, columns, boxes, and distinct-digit cages.

When all other digits have been eliminated from a cell, the remaining candidate is the
cell's value by exhaustion. Once placed, the digit cannot appear in any peer cell that
shares a unit with it.

Proof: Let C have candidates = {d}. Every other digit d' ≠ d has already been eliminated
from C. Therefore C = d. For any peer P sharing a unit U with C: if d ∈ candidates(P),
the unit constraint would require two occurrences of d in U — contradiction. Therefore d
can be eliminated from every such P.

Guards:
  ctx.cell !== null      engine sets this only when CELL_DETERMINED fires
  ctx.hintDigit !== null engine sets this to the sole remaining candidate`.trim();
  readonly priority = 0;
  readonly triggers: ReadonlySet<Trigger> = new Set([Trigger.CELL_DETERMINED]);
  readonly unitKinds: ReadonlySet<UnitKind> = new Set();

  apply(ctx: RuleContext): RuleResult {
    if (ctx.cell === null || ctx.hintDigit === null) return emptyResult();
    const [r, c] = ctx.cell;
    const d = ctx.hintDigit;
    return { ...emptyResult(), placements: [{ cell: ctx.cell, digit: d }], eliminations: ctx.board.peerEliminations(r, c, d) };
  }

  asHints(ctx: RuleContext, eliminations: readonly Elimination[]): HintResult[] {
    if (ctx.cell === null || ctx.hintDigit === null) return [];
    const [r, c] = ctx.cell;
    const d = ctx.hintDigit;
    const peerNote = eliminations.length > 0
      ? ` This also removes ${d} from ${eliminations.length === 1 ? '1 peer' : `${eliminations.length} peers`}: ${eliminations.map(e => cellLabel(e.cell)).join(', ')}.`
      : '';
    return [{
      ruleName: this.name,
      displayName: 'Naked Single',
      explanation: `Cell ${cellLabel([r, c] as Cell)} has only one remaining candidate: ${d}. Place ${d} there.${peerNote}`,
      highlightCells: [[r, c] as Cell],
      eliminations,
      placement: [r, c, d],
      virtualCageSuggestion: null,
    }];
  }
}
